File: packages/typescript/src/resources/suppressions.ts
/**
 * The `suppressions` resource: numbers that must never receive messages.
 * @module
 */
import { toWire } from '../models.js';
import type {
  CreateSuppressionParams,
  ListParams,
  Page,
  RequestOptions,
  Suppression,
  SuppressionImportResult,
} from '../models.js';
import { Resource, seg, wireQuery } from './base.js';

/** Accessed as `client.suppressions`. */
export class Suppressions extends Resource {
  list(params: ListParams = {}, options?: RequestOptions): Promise<Page<Suppression>> {
    return this.page({ method: 'GET', path: '/v1/compliance/suppressions', query: wireQuery(params) }, options);
  }

  /** Suppress one number. Never retried. */
  create(params: CreateSuppressionParams, options?: RequestOptions): Promise<Suppression> {
    return this.one({ method: 'POST', path: '/v1/compliance/suppressions', json: toWire(params) }, options);
  }

  /** Suppress many numbers at once. Never retried. */
  import(items: CreateSuppressionParams[], options?: RequestOptions): Promise<SuppressionImportResult> {
    return this.one(
      { method: 'POST', path: '/v1/compliance/suppressions/import', json: toWire({ items }) },
      options,
    );
  }

  delete(id: number | string, options?: RequestOptions): Promise<void> {
    return this.none({ method: 'DELETE', path: `/v1/compliance/suppressions/${seg(id)}` }, options);
  }
}

File: packages/rust/src/resources/suppressions.rs
//! The `suppressions` resource: the do-not-send list.

use serde::Serialize;

use super::list_query;
use crate::error::Result;
use crate::http::{HttpClient, Idem};
use crate::models::{CreateSuppression, ListParams, Page, Suppression, SuppressionImportResult};

#[derive(Serialize)]
struct ImportBody<'a> {
    items: &'a [CreateSuppression],
}

/// Accessed as `client.suppressions()`.
#[derive(Debug, Clone)]
pub struct Suppressions {
    http: HttpClient,
}

impl Suppressions {
    pub(crate) fn new(http: HttpClient) -> Self {
        Self { http }
    }

    /// List suppressions.
    pub async fn list(&self, params: ListParams) -> Result<Page<Suppression>> {
        self.http
            .get(&format!(
                "/v1/compliance/suppressions{}",
                list_query(&params)
            ))
            .await
    }

    /// Suppress a number. Not retried.
    pub async fn create(&self, params: &CreateSuppression) -> Result<Suppression> {
        self.http
            .send("POST", "/v1/compliance/suppressions", params, Idem::None)
            .await
    }

    /// Bulk-import suppressions. Not retried.
    pub async fn import(&self, items: &[CreateSuppression]) -> Result<SuppressionImportResult> {
        self.http
            .send(
                "POST",
                "/v1/compliance/suppressions/import",
                &ImportBody { items },
                Idem::None,
            )
            .await
    }

    /// Remove a suppression.
    pub async fn delete(&self, id: i64) -> Result<()> {
        self.http
            .empty(
                "DELETE",
                &format!("/v1/compliance/suppressions/{id}"),
                Idem::None,
            )
            .await
    }
}
